gph/src/ripser.h

```diff
@@ -1432,48 +1435,71 @@ class ripser
                                            // the new pair; could elide an
                                            // extra atomic load
 
-                        /* Pairs should be extracted if insertion was
-                         * first one ! */
-                        size_t location = last_diameter_index++;
-                        diameters[location] = get_diameter(pivot);
-                        auto first_ins =
-                            deaths
-                                .insert({get_index(get_entry(pivot)), location})
-                                .second;
-
-                        /* Only insert when it is the first time this bar is
-                         * encountered
-                         */
-                        if (return_flag_persistence_generators && first_ins) {
-                            // Inessential flag persistence generators in dim >
-                            // 0
-                            std::vector<index_t> vertices_birth(dim + 1);
-                            std::vector<index_t> vertices_death(dim + 2);
-
-                            index_t birth_idx =
-                                get_index(get_entry(column_to_reduce));
-                            index_t death_idx = get_index(get_entry(pivot));
-
-                            get_simplex_vertices(birth_idx, dim, n,
-                                                 vertices_birth.rbegin());
-                            get_simplex_vertices(death_idx, dim + 1, n,
-                                                 vertices_death.rbegin());
-
-                            edge_t birth_edge =
-                                get_youngest_edge_simplex(vertices_birth);
-                            edge_t death_edge =
-                                get_youngest_edge_simplex(vertices_death);
-
-                            finite_generator[location] = {
-                                birth_edge.first, birth_edge.second,
-                                death_edge.first, death_edge.second};
+                        // Infinite death indicates an essential bar in disguise
+                        value_t death = get_diameter(pivot);
+                        is_essential =
+                            (death == std::numeric_limits<value_t>::infinity());
+                        if (!is_essential) {
+                            /* Pairs should only be extracted if insertion was
+                             * first one! */
+                            size_t location = last_diameter_index++;
+                            diameters[location] = death;
+                            auto first_ins =
+                                deaths
+                                    .insert({get_index(get_entry(pivot)), location})
+                                    .second;
```

Comment:
I don't yet understand these lines too well so please check that them now being inside this if cannot lead to wrong results.

Reply:
It's an explanation of what is going on. Because the reduction is done in parallel. You can have at a certain point 2 differents workers getting the same `pivot` as the last reduction. If that happens, we ensure with this check that only one barcode is produced instead of two.

I think that the comment should remain for anyone looking at the code

Reply:
Sure, I wasn't complaining about the comment, just wondering about correctness of the code in this part especially.

Reply:
Oh !
Is it more clear with my previous explanation ?